src/bitcoind/interface.rs

```diff
@@ -89,64 +86,48 @@ impl BitcoinD {
             node_client,
             watchonly_client,
             cpfp_client,
+            retries: 0,
         })
     }
 
-    // Reasonably try to be robust to possible spurious communication error.
-    fn handle_error(&self, e: jsonrpc::Error, start: Instant) -> Result<(), BitcoindError> {
-        let now = Instant::now();
-
-        match e {
-            jsonrpc::Error::Transport(ref err) => {
-                log::error!("Transport error when talking to bitcoind: '{}'", err);
-
-                // This is *always* a simple_http::Error. Rule out the error that can
-                // not occur after startup (ie if we encounter them it must be startup
-                // and we better be failing quickly).
-                let any_err = err as &dyn Any;
-                if let Some(http_err) = any_err.downcast_ref::<HttpError>() {
-                    match http_err {
-                        HttpError::InvalidUrl { .. } => return Err(BitcoindError::Server(e)),
-                        // FIXME: allow it to be unreachable for a handful of seconds,
-                        // but not at startup!
-                        HttpError::SocketError(_) => return Err(BitcoindError::Server(e)),
-                        HttpError::HttpParseError => {
-                            // Weird. Try again once, just in case.
-                            if now.duration_since(start) > Duration::from_secs(1) {
-                                return Err(BitcoindError::Server(e));
+    /// Set the retry limit field limiting the number of retry a client is allowed to do for a request.
+    pub fn with_retry_limit(mut self, retry_limit: usize) -> Self {
+        self.retries = retry_limit;
+        self
+    }
+
+    /// Wrapper to retry a request sent to bitcoind upon IO failure
+    /// according to the configured number of retries.
+    fn retry<T, R: Fn() -> Result<T, BitcoindError>>(
+        &self,
+        request: R,
+    ) -> Result<T, BitcoindError> {
+        let mut error: Option<BitcoindError> = None;
+        for i in 0..self.retries + 1 {
+            match request() {
+                Ok(res) => return Ok(res),
+                Err(e) => match e {
+                    BitcoindError::Server(jsonrpc::Error::Transport(ref err)) => {
+                        match err.downcast_ref::<HttpError>() {
+                            Some(HttpError::Timeout)
+                            | Some(HttpError::SocketError(_))
+                            | Some(HttpError::HttpErrorCode(503)) => {
```

Comment:
Hmm. I think we should retry on warmup (for instance if bitcoind was restarted from under us). It's RPC error `-28`. Is there anything else we'd like to cover? Maybe an http `500`? Not sure, it was used before 22.0 for limit exceeded but we don't support previous versions anyways. Looks like it's used to send JSONRPC errors too. Apart from warmup i think we want to let JSONRPC error go through right?

Reply:
To sum things up we want:
1. fail fast at startup EXCEPT when the error is the warming up error: then it must retry
2. retry only timeout, io, and too much requests errors

For the first, because at startup the retry count is set at 0, I did not handle the warming up error and I let the current behavior: https://github.com/revault/revaultd/blob/master/src/bitcoind/mod.rs#L136
but I can modify the client to loop the retry without limit for the warming up error.
What do you think ?

Reply:
I think we want to:
1. At startup, never retry but when encountering a warmup error
2. During execution, retry on any non-fatal error that may occur (i think warmup is one case we need to handle here)

(1) we already do. We just need to do (2) which AFAICT is only missing the warmup special case (if i'm not missing anything.
https://github.com/revault/revaultd/blob/457ff85dc40834ccf9e78ced25547a77d7278ad5/src/bitcoind/mod.rs#L135-L142